pages/blog/the-reference-rabbit-hole.md

```diff
@@ -0,0 +1,139 @@
+
+```yaml
+asyncapi: '2.2.0'
+info:
+  title: Test overriding dereferenced objects 
+  version: '1.0.0'
+servers:
+  test-server:
+    $ref: './servers/testServer.yaml'
+...
+```
+
+Together with `testServer.yaml`:
+```yaml
+url: ws://mycompany.com/ws
+protocol: ws
+```
+
+Validating the AsyncAPI document using the [JSON Schema representation for 2.2.0](https://github.com/asyncapi/spec-json-schemas/blob/master/schemas/2.2.0.json), would reject it. 
+
+However, because the parser bundles the reference together before doing so:
+
+```yaml
+asyncapi: '2.2.0'
+info:
+  title: Test overriding dereferenced objects 
+  version: '1.0.0'
+servers:
+  test-server:
+    url: ws://mycompany.com/ws
+    protocol: ws
+...
+```
+
+It is a fully valid AsyncAPI document once the parser tries to validate the document. Checkout [parser-js #405](https://github.com/asyncapi/parser-js/issues/405) for more information.
+
+## Deep diving
+We started discussing the reference behavior, and we started to peruse the specification in terms of the Reference Object and the [Schema Object](https://www.asyncapi.com/docs/specifications/v2.2.0#schemaObject).
+
+That was when we took notice, that depending on where you use the `$ref` keyword, it is subject to different behavior.
+
+This was something that, I especially, did not realize up until this point. In AsyncAPI Reference Object, is following [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03), and for our Schema Object something different...
+
+By default, `$ref` in the Schema Object, is inherited by [JSON Schema Draft 7](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01), described in [section 8](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8).
+
+One of the key differences between our Reference Object, and how `$ref` is dereferenced in JSON Schema Draft 7, is the [$id keyword](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-8.2). This allows you to define a URI that other schemas within itself need to resolve against. Meaning that it is allowed to do:
+
+```yaml
+asyncapi: "2.2.0"
+...
+channels:
+  test/channel:
+    publish:
+      message:
+        payload: 
+          $id: "https://example.com/schemas/customer"
+          type: "object"
+          properties:
+            address: 
+              $ref: "/schemas/address"
+components: 
+  schemas:
+    # Same dereference behavior
+...
+```
+
+This results in the dereferencing logic to look up the reference for the `address` property, at `https://example.com/schemas/address`, because it uses the Base URI in `$id` from the parent schema (`https://example.com`). I tried a little test in the [new Studio](https://studio.asyncapi.com/), which showed that this was not supported by the parser. See [parser-js #403](https://github.com/asyncapi/parser-js/issues/403) for more information.
```

Comment:
nitpick

```suggestion
This results in the dereferencing logic to look up the reference for the `address` property at `https://example.com/schemas/address` because it uses the Base URI in `$id` from the parent schema (`https://example.com`). I tried a little test in the [new Studio](https://studio.asyncapi.com/), which showed that this was not supported by the parser. See [parser-js #403](https://github.com/asyncapi/parser-js/issues/403) for more information.
```